### PR TITLE
Add Kernel for LagLeadOperator & RollingWindowFeaturizer

### PR DESCRIPTION
**Pull request by wangyems**
Add ORT Kernel for LagLeadOperator
Add ORT Kernel for RollingWindowFeaturizer
keep FeaturizerLibrary commit up to date
fix test cases in ShortGrainDropper kernel

### CODE REVIEW

File: onnxruntime/core/graph/featurizers_ops/featurizers_defs.cc [hunk 1]
```diff
@@ -674,6 +678,132 @@ void RegisterLabelEncoderFeaturizerVer1() {
...
+      // [value[offset[K - 1] - N], value[offset[K - 1] - (N - 1)], ..., value[offset[K - 1] - 1]]
+      // ]
+
+      // The resulting matrix size is K rows x N cols, where K is the number of offsets and N is the horizon.
+
+      // Horizon and offsets should be passed in during construction. Offsets are passed in as a vector of ints so multiple lag orders can be applied within one featurizer call.
+      // Output type is a tuple of vector of string, which representing grains, and a matrix. The matrix is of optional<T> where rows are grouped by different offsets and columns are grouped by horizon.
+
+      // C++-style pseudo signature:
+      //   template <typename T> tuple<vector<string>,matrix<T?>> execute(std::vector<std::string> const &, T const &> const &value);
```
Comment (yuslepukhin) [inline onnxruntime/core/graph/featurizers_ops/featurizers_defs.cc:700]: >T? [](start = 61, length = 2)

This is not Java

File: onnxruntime/core/graph/featurizers_ops/featurizers_defs.cc [hunk 2]
```diff
@@ -674,6 +678,132 @@ void RegisterLabelEncoderFeaturizerVer1() {
...
+      .TypeAndShapeInferenceFunction(
+          [](ONNX_NAMESPACE::InferenceContext& ctx) {
+            propagateElemTypeFromInputToOutput(ctx, 1, 0);
+            auto input_elem_type = ctx.getInputType(2)->tensor_type().elem_type();
+            if (input_elem_type == ONNX_NAMESPACE::TensorProto_DataType_FLOAT) {
+              propagateElemTypeFromDtypeToOutput(ctx, ONNX_NAMESPACE::TensorProto_DataType_FLOAT, 1);
+            } else if (input_elem_type == ONNX_NAMESPACE::TensorProto_DataType_DOUBLE) {
+              propagateElemTypeFromDtypeToOutput(ctx, ONNX_NAMESPACE::TensorProto_DataType_DOUBLE, 1);
+            } else {
+              fail_type_inference("input 2 is expected to have an accepted type");
```
Comment (yuslepukhin) [inline onnxruntime/core/graph/featurizers_ops/featurizers_defs.cc:784]: >fail_type_inference("input 2 is expected to have an accepted type"); [](start = 14, length = 68)

The constraints already make sure that the type is acceptable. Just propagate the type.

File: onnxruntime/core/graph/featurizers_ops/featurizers_defs.cc [hunk 3]
```diff
@@ -674,6 +678,132 @@ void RegisterLabelEncoderFeaturizerVer1() {
...
+            }
+            if (hasInputShape(ctx, 1)) {
+              const auto& grains_shape = getInputShape(ctx, 1);
+              if (grains_shape.dim_size() != 2) {
+                fail_shape_inference("Expecting Grains to have 2 dimensions");
+              }
+              propagateShapeFromInputToOutput(ctx, 1, 0);
```
Comment (yuslepukhin) [inline onnxruntime/core/graph/featurizers_ops/featurizers_defs.cc:791]: >propagateShapeFromInputToOutput(ctx, 1, 0); [](start = 14, length = 43)

This is not needed, I believe.

File: onnxruntime/featurizers_ops/cpu/laglead_operator_transformer.cc [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+    // Prepare the Output
+    const int64_t output_dim_0 = grains_tensor->Shape()[0];
+
+    T* output_data;
+    bool has_allocate_output_data = false;
+    std::function<void(OutputType)> callback_fn;
+    callback_fn = [ctx, &output_grains_data, &output_data, &has_allocate_output_data, output_dim_0](OutputType value) -> void {
+      GrainT const output_grains(std::get<GrainT>(value));
+      OutputMatrixType const output_matrix(std::get<OutputMatrixType>(value));
```
Comment (yuslepukhin) [inline onnxruntime/featurizers_ops/cpu/laglead_operator_transformer.cc:56]: >utputMatrixType [](start = 7, length = 15)

You are creating a copy of the matrix but all you need is a reference.

File: onnxruntime/featurizers_ops/cpu/laglead_operator_transformer.cc [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+      //Allocate tensor memory after first output is generated
+      if (!has_allocate_output_data) {
+        TensorShape output_shape({output_dim_0, output_matrix.rows(), output_matrix.cols()});
+        Tensor* output_tensor(ctx->Output(1, output_shape));
+        output_data = output_tensor->MutableData<T>();
+        has_allocate_output_data = true;
+      }
+      output_grains_data = std::copy(output_grains.begin(), output_grains.end(), output_grains_data);
+      output_data = std::copy(output_matrix.data(), output_matrix.data() + output_matrix.size(), output_data);
```
Comment (yuslepukhin) [inline onnxruntime/featurizers_ops/cpu/laglead_operator_transformer.cc:65]: >output_data = std::copy(output_matrix.data(), output_matrix.data() + output_matrix.size(), output_ [](start = 6, length = 98)

You should use matrix mapping, see other kernels for examples.

File: onnxruntime/featurizers_ops/cpu/rolling_window_transformer.cc [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+    std::function<void(OutputType)> callback_fn;
+    callback_fn = [ctx, &output_data, &has_allocate_output_data, output_dim_0](OutputType value) -> void {
+      //Allocate tensor memory after first output is generated
+      if(!has_allocate_output_data) {
+        TensorShape output_shape({output_dim_0, static_cast<int64_t>(value.size())});
+        Tensor* output_tensor(ctx->Output(0, output_shape));
+        output_data = output_tensor->MutableData<double>();
+        has_allocate_output_data = true;
+      }
+      output_data = std::copy(value.data(), value.data() + value.size(), output_data);
```
Comment (yuslepukhin) [inline onnxruntime/featurizers_ops/cpu/rolling_window_transformer.cc:53]: > output_data = std::copy(value.data(), value.data() + value.size(), output_data); [](start = 5, length = 81)

Same comments as previous

File: onnxruntime/featurizers_ops/cpu/rolling_window_transformer.cc [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+    const auto grains_num = grains_tensor->Shape()[1];
+
+    // Get the Target
+    const auto* target_tensor(ctx->Input<Tensor>(2));
+    const T* target_data(target_tensor->Data<T>());
+
+    // Prepare the output
+    const auto output_dim_0 = grains_tensor->Shape()[0];
+
+    double* output_data(0);
```
Comment (yuslepukhin) [inline onnxruntime/featurizers_ops/cpu/rolling_window_transformer.cc:42]: >double* output_data(0); [](start = 4, length = 23)

= nullptr

**Comment by yuslepukhin** [comment]
This PR is about laglead and rolling_window kernels. How come short_grain is being updated here? I think it should be a separate PR>

File: onnxruntime/featurizers_ops/cpu/laglead_operator_transformer.cc (hunk 5 above)
Comment (wangyems) [reply onnxruntime/featurizers_ops/cpu/laglead_operator_transformer.cc:65]: we discuss in the scrum that it's better to keep as it is this time and will support matrix mapping type in FeaturizersLibrary in the future. I leave a todo comment there

**Pull request closed without merging**